libraid/bech32/bech32.go

```diff
@@ -0,0 +1,217 @@
+		return "", fmt.Errorf("invalid hrp : hrp=%v", hrp)
+	}
+	for p, c := range hrp {
+		if c < 33 || c > 126 {
+			return "", fmt.Errorf("invalid character human-readable part : hrp[%d]=%d", p, c)
+		}
+	}
+	if strings.ToUpper(hrp) != hrp && strings.ToLower(hrp) != hrp {
+		return "", fmt.Errorf("mix case : hrp=%v", hrp)
+	}
+	lower := strings.ToLower(hrp) == hrp
+	hrp = strings.ToLower(hrp)
+	combined := append(data, createChecksum(hrp, data)...)
+	var ret bytes.Buffer
+	ret.WriteString(hrp)
+	ret.WriteString("1")
+	for idx, p := range combined {
+		if p < 0 || p >= len(charset) {
+			return "", fmt.Errorf("invalid data : data[%d]=%d", idx, p)
+		}
+		ret.WriteByte(charset[p])
+	}
+	if lower {
+		return ret.String(), nil
+	}
+	return strings.ToUpper(ret.String()), nil
+}
+
+// Decode decodes bechString(Bech32) returns hrp(human-readable part) and data(32bit data array) / or error
+func Decode(bechString string) (string, []int, error) {
+	if len(bechString) > 90 {
+		return "", nil, fmt.Errorf("too long : len=%d", len(bechString))
+	}
+	if strings.ToLower(bechString) != bechString && strings.ToUpper(bechString) != bechString {
+		return "", nil, fmt.Errorf("mixed case")
+	}
+	bechString = strings.ToLower(bechString)
+	pos := strings.LastIndex(bechString, "1")
+	if pos < 1 || pos+7 > len(bechString) {
+		return "", nil, fmt.Errorf("separator '1' at invalid position : pos=%d , len=%d", pos, len(bechString))
+	}
+	hrp := bechString[0:pos]
+	for p, c := range hrp {
+		if c < 33 || c > 126 {
+			return "", nil, fmt.Errorf("invalid character human-readable part : bechString[%d]=%d", p, c)
+		}
+	}
+	data := []int{}
+	for p := pos + 1; p < len(bechString); p++ {
+		d := strings.Index(charset, fmt.Sprintf("%c", bechString[p]))
+		if d == -1 {
+			return "", nil, fmt.Errorf("invalid character data part : bechString[%d]=%d", p, bechString[p])
+		}
+		data = append(data, d)
+	}
+	if !verifyChecksum(hrp, data) {
+		return "", nil, fmt.Errorf("invalid checksum")
+	}
+	return hrp, data[:len(data)-6], nil
+}
+
+func convertbits(data []int, frombits, tobits uint, pad bool) ([]int, error) {
+	acc := 0
+	bits := uint(0)
+	ret := []int{}
+	maxv := (1 << tobits) - 1
+	for idx, value := range data {
+		if value < 0 || (value>>frombits) != 0 {
+			return nil, fmt.Errorf("invalid data range : data[%d]=%d (frombits=%d)", idx, value, frombits)
+		}
+		acc = (acc << frombits) | value
+		bits += frombits
+		for bits >= tobits {
+			bits -= tobits
+			ret = append(ret, (acc>>bits)&maxv)
+		}
+	}
+	if pad {
+		if bits > 0 {
+			ret = append(ret, (acc<<(tobits-bits))&maxv)
+		}
+	} else if bits >= frombits {
+		return nil, fmt.Errorf("illegal zero padding")
+	} else if ((acc << (tobits - bits)) & maxv) != 0 {
+		return nil, fmt.Errorf("non-zero padding")
+	}
+	return ret, nil
+}
+
+// SegwitAddrDecode decodes hrp(human-readable part) Segwit Address(string), returns version(int) and data(bytes array) / or error
+func SegwitAddrDecode(hrp, addr string) (int, []int, error) {
```

Comment:
do we actually need this?

Reply:
yes, the Intent Identifier is generated by this, and this is the reason I need pull in this code, otherwise existing bash32 lib in golang is good enough, I can rewrite it but seems not worth it (technically I did once, and deleted later)